lib/deno/mod.ts

```diff
@@ -178,63 +178,142 @@ interface Service {
 
 let defaultWD = Deno.cwd()
 let longLivedService: Promise<Service> | undefined
-let stopService: (() => void) | undefined
+let stopService: (() => Promise<void>) | undefined
+
+// Declare a common subprocess API for the two implementations below
+type SpawnFn = (cmd: string, options: {
+  args: string[]
+  stdin: 'piped' | 'inherit'
+  stdout: 'piped' | 'inherit'
+  stderr: 'inherit'
+}) => {
+  stdin: {
+    write(bytes: Uint8Array): void
+    close(): void
+  }
+  stdout: {
+    read(): Promise<Uint8Array | null>
+    close(): void
+  }
+  close(): Promise<void> | void
+  status(): Promise<{ code: number }>
+}
+
+// Deno ≥1.40
+const spawnNew: SpawnFn = (cmd, { args, stdin, stdout, stderr }) => {
+  const child = new Deno.Command(cmd, {
+    args,
+    cwd: defaultWD,
+    stdin,
+    stdout,
+    stderr,
+  }).spawn()
+  const writer = child.stdin.getWriter()
+  const reader = child.stdout.getReader()
+  return {
+    stdin: {
+      write: bytes => writer.write(bytes),
+      close: () => writer.close(),
+    },
+    stdout: {
+      read: () => reader.read().then(x => x.value || null),
+      close: () => reader.cancel(),
+    },
+    close: async () => {
+      // Note: This can throw with EPERM on Windows (happens in GitHub Actions)
+      child.kill()
```

Comment:
Trying to kill the child process on Windows throws `PermissionDenied: Access is denied. (os error 5)`. But not killing the process on Windows causes Deno to fail tests with `1 async operation to op_spawn_wait was started in this test, but never completed`. So it seems like maybe there's no way to get tests to pass on Windows with Deno's new APIs? I'm not sure how to proceed here. Maybe this is a bug in Deno itself?

Reply:
in `stopService` you may need to await `stdin.close()` since `writer.close()` returns a promise? (same for `stdout` maybe though [Deno docs](https://deno.land/api@v1.40.1?s=Deno.Command) say only the "stdin `WritableStream` needs to be closed manually") - sorry I don't have Windows on hand to test.

I think the first test fail because the child process is spawned within `fn` but closed outside in `finally { esbuildNative.stop() }`. Seems like Deno test is being overly critical/protective there as it's working as intended. I don't know enough to resolve that issue sadly.

Reply:
This looks like a race between you calling `stdin.close()` and `child.kill()`. The esbuild binary probably exits by itself when you close it's stdin to prevent runaway processes. Thus, when you call `stdin.close()` the process may have already exited by the time you call `child.kill()` (and then that call fails because the process is already gone).

You can rely on the fact that esbuild shuts down when you close `stdin`: just make sure you close `stdin` during shutdown, and then just wait for the process to actually have exited with `await child.status`.

Reply:
That seems to have worked. Thanks for the suggestion!